File: backend/app/schemas/assets.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator


AssetType = Literal["character", "item", "tile", "background", "ui"]
AssetStyle = Literal["pixel", "cartoon", "ink", "dark", "sci-fi"]
PaletteName = Literal["ember", "forest", "ocean", "candy", "mono"]
QualityStatus = Literal["pass", "warn", "fail"]


class GenerateRequest(BaseModel):
    project_name: str = Field(default="Untitled Game", min_length=1, max_length=80)
    description: str = Field(min_length=1, max_length=500)
    asset_type: AssetType = "item"
    style: AssetStyle = "pixel"
    size: int = Field(default=128, ge=32, le=512)
    count: int = Field(default=4, ge=1, le=8)
    transparent_background: bool = True
    palette: PaletteName = "ocean"
    consistency_seed: str = Field(default="default", min_length=1, max_length=80)
    style_pack_name: str | None = Field(default=None, max_length=80)

    @field_validator("size")
    @classmethod
    def size_must_be_power_of_two(cls, value: int) -> int:
        if value not in {32, 64, 128, 256, 512}:
            raise ValueError("size must be one of 32, 64, 128, 256, 512")
        return value

    @field_validator("project_name", "description", "consistency_seed")
    @classmethod
    def trim_text(cls, value: str) -> str:
        return value.strip()

    @field_validator("style_pack_name")
    @classmethod
    def trim_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        return trimmed or None


class GeneratedAsset(BaseModel):
    id: str
    name: str
    asset_type: AssetType
    style: AssetStyle
    size: int
    data_url: str
    seed: str
    prompt: str
    created_at: str
    metadata: dict[str, str | int | bool]


class QualityCheck(BaseModel):
    key: str
    label: str
    status: QualityStatus
    detail: str


class GenerateResponse(BaseModel):
    request_id: str
    enhanced_prompt: str
    constraints: list[str]
    quality_checks: list[QualityCheck]
    assets: list[GeneratedAsset]

File: frontend/src/types/assets.ts
export type AssetType = 'character' | 'item' | 'tile' | 'background' | 'ui'

export type AssetStyle = 'pixel' | 'cartoon' | 'ink' | 'dark' | 'sci-fi'

export type PaletteName = 'ember' | 'forest' | 'ocean' | 'candy' | 'mono'

export type QualityStatus = 'pass' | 'warn' | 'fail'

export interface GenerateRequest {
  projectName: string
  description: string
  assetType: AssetType
  style: AssetStyle
  size: number
  count: number
  transparentBackground: boolean
  palette: PaletteName
  consistencySeed: string
  stylePackName?: string
}

export interface GeneratedAsset {
  id: string
  name: string
  assetType: AssetType
  style: AssetStyle
  size: number
  dataUrl: string
  seed: string
  prompt: string
  createdAt: string
  metadata: Record<string, string | number | boolean>
}

export interface GenerateResponse {
  requestId: string
  enhancedPrompt: string
  constraints: string[]
  qualityChecks: QualityCheck[]
  assets: GeneratedAsset[]
}

export interface QualityCheck {
  key: string
  label: string
  status: QualityStatus
  detail: string
}

export interface StylePack {
  id: string
  name: string
  description: string
  assetType: AssetType
  style: AssetStyle
  size: number
  transparentBackground: boolean
  palette: PaletteName
  consistencySeed: string
  updatedAt: string
}

export interface StylePreset {
  label: string
  value: AssetStyle
  hint: string
}

export interface AssetTypePreset {
  label: string
  value: AssetType
  hint: string
}

export interface PalettePreset {
  label: string
  value: PaletteName
  colors: string[]
}
